lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp

```diff
@@ -0,0 +1,306 @@
+  auto mem_buffer_up = index_cache->GetCachedData(module_name);
+
+  // Nothing cached.
+  if (!mem_buffer_up) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] No cached file found for module {0}.",
+              module->GetFileSpec().GetFilename());
+    return;
+  }
+
+  // Extractor used to extract the header information (see the .h file for
+  // details on the format).
+  DataExtractor header_extractor(mem_buffer_up->getBufferStart(),
+                                 mem_buffer_up->getBufferSize(),
+                                 module->GetObjectFile()->GetByteOrder(),
+                                 module->GetObjectFile()->GetAddressByteSize());
+
+  lldb::offset_t read_offset = 0;
+  // The version of the cache format this was encoded with.
+  uint16_t cache_version = 0;
+  if (!header_extractor.GetU16(&read_offset, &cache_version, 1)) {
+    LLDB_LOG(log,
+             "[SwiftMetadataCache] Failed to read cache file version for {0}.",
+             module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  // It only makes sense to process caches with the same version.
+  // Delete the cache so we can recreate it later on otherwise.
+  if (cache_version != SWIFT_METADATA_CACHE_VERSION) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] Cache file version mismatch for {0}.",
+              module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  std::string UUID = module->GetUUID().GetAsString();
+  std::string cached_UUID = header_extractor.GetCStr(&read_offset);
+  // If no uuid in the file something is wrong with the cache.
+  if (cached_UUID.empty()) {
+    LLDB_LOG(log,
+             "[SwiftMetadataCache] Failed to read cached UUID for module {0}.",
+             module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  // If the UUIDs don't match this is most likely a stale cache.
+  if (cached_UUID != UUID) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] Module UUID mismatch for {0}.",
+              module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  uint64_t expanded_size = 0;
+  if (!header_extractor.GetU64(&read_offset, &expanded_size, 1)) {
+    LLDB_LOGV(log,
+              "[SwiftMetadataCache] Failed to read decompressed cache size for "
+              "module {0}.",
+              module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  const auto *start =
+      (const char *)header_extractor.GetData(&read_offset, 0);
+  // Create a reference to the compressed data.
+  llvm::StringRef string_buffer(start, (uint64_t)mem_buffer_up->getBufferEnd() -
+                                           (uint64_t)start);
+
+  llvm::SmallString<0> decompressed;
+  auto error =
+      llvm::zlib::uncompress(string_buffer, decompressed, expanded_size);
```

Comment:
For consideration: Using a format that needs to be decompressed is more expensive in terms of RAM than mmapping in data that we can use directly without dirtying any memory. But it would use more disk space.

Reply:
Does LLVM's OnDiskHashTable work without copying the working set into RAM?

Reply:
I think that's a good idea, but we should try to nail the format in a second patch, as this one is pretty big already.